Enum memory representation and well-formedness (#146)
minirust/minirust: spec/lang/representation.md

### Enums

Enum encoding and decoding.
Note that the discriminant may not be written into bytes that contain encoded data.
This is to ensure that pointers to the data always contain valid values.

```rust

> /// Computes the discriminant from the value's bytes by the decision tree in the `Discriminator` by
> /// recursing over the branches defined by `Discriminator::Unknown`.

Reviewer: ```suggestion
/// Reads the discriminant from the representation bytes, using the given `Discriminator`.
```

/// FIXME: we have multiple quite different fail sources, it would be nice to return more error information.

> fn compute_discriminant<M: Memory>(bytes: List<AbstractByte<M::Provenance>>, discriminator: Discriminator) -> Option<Int> {

Reviewer: "compute" is not entirely accurate, we're not doing much computation here after all. I think this should be called "decode_discriminant"; that also makes it clear which operation this is a helper for.

    match discriminator {
        Discriminator::Known(val) => Some(val),
        Discriminator::Invalid => None,
        Discriminator::Unknown { offset, children, fallback } => {
            let AbstractByte::Init(val, _) = bytes[offset.bytes()]
                else { return None };
            let next_discriminator = children.get(val).unwrap_or(fallback);
            compute_discriminant::<M>(bytes, next_discriminator)
        }
    }
}

impl Type {
    fn decode<M: Memory>(Type::Enum { variants, discriminator, size, .. }: Self, bytes: List<AbstractByte<M::Provenance>>) -> Option<Value<M>> {
        if bytes.len() != size.bytes() { throw!(); }
        let disc = compute_discriminant::<M>(bytes, discriminator)?;

        // Decode into the variant.
        // Because the variant is the same size as the enum we don't need to pass a subslice.
        let Some(value) = variants[disc].ty.decode(bytes)
            else { return None };

        Some(Value::Variant { idx: disc, data: value })
    }

    fn encode<M: Memory>(Type::Enum { variants, .. }: Self, val: Value<M>) -> List<AbstractByte<M::Provenance>> {
        let Value::Variant { idx, data } = val else { panic!() };

>         // idx is to be guaranteed in bounds by the well-formed check in the type.

Reviewer: ```suggestion
        // `idx` is guaranteed to be in bounds by the well-formed check in the type.
```

        let Variant { ty: variant, tagger } = variants[idx];
        let mut bytes = variant.encode(data.extract());

>         // write tag afterwards into the encoded bytes. This is fine as we don't allow

Reviewer: ```suggestion
        // Write tag into the encoded bytes. This is fine as we don't allow
```

        for (offset, value) in tagger.iter() {

>             assert_eq!(Some(AbstractByte::Uninit), bytes.get(offset.bytes()));

Reviewer: This assertion is somewhat dubious. We should have the invariant that "if your code passes well-formedness then you will never get a panic", but here, this panic is not excluded by the WF-check.

So I think this should be just a comment.
```suggestion
            // If this enum type is "reasonable", we'll only overwrite `Uninit` here`.
            // However check_wf cannot ensure this, so we do not have an assertion.
```

### Enums

Enum encoding and decoding.
Note that the discriminant setting and reading does not modify any bytes.
This is does not allow types like `Result<bool, bool>` to be stored in one byte.
However this is fine as Rust currently does not do any optimizations like that

> and doing it here will introduce more questions.

Reviewer: I don't understand this paragraph. How does it nit modify any bytes...?

`Result<bool, bool>` can't be stored in one byte since the types have overlapping valid byte ranges.

```rust

> fn compute_discriminant<M: Memory>(bytes: List<AbstractByte<M::Provenance>>, discriminator: Discriminator) -> Option<Int> {

Reviewer: What is this function doing? Please add a doc comment.

>     let mut disc = discriminator;

Reviewer: Please use recursion instead of mutable state with a loop.

    loop {
        match disc {
            Discriminator::Known(val) => break Some(val),
            Discriminator::Unknown { offset, children } => {

>                 if offset < 0 || offset >= bytes.len() { throw!(); }

Reviewer: This should be checked already by well-formedness.

>                 let AbstractByte::Init(val, _) = bytes.get(offset).unwrap()

Reviewer: ```suggestion
                let AbstractByte::Init(val, _) = bytes[offset]
```
does this not work?

                    else { break None };
                    // else { throw_ub!("Encountered uninitialized byte while computing enum discriminant.") };
                let Some(new_disc) = children.get(val)
                    else { break None };
                    // else { throw_ub!("Encountered invalid discriminant value.") };
                disc = new_disc;
            }
        }
    }
}

impl Type {
    fn decode<M: Memory>(Type::Enum { variants, discriminator, size, .. }: Self, bytes: List<AbstractByte<M::Provenance>>) -> Option<Value<M>> {
        if bytes.len() != size.bytes() { throw!(); }
        let disc = compute_discriminant::<M>(bytes, discriminator)?;

        // The discriminator does the discriminant check and as such should not be able to return an invalid value.

>         if disc < Int::ZERO || disc >= variants.len() { throw!(); }

Reviewer: This should not be needed, well-formedness should check this.

        variants[disc].inner.decode(bytes)
    }

    fn encode<M: Memory>(Type::Enum { variants, size, .. }: Self, val: Value<M>) -> List<AbstractByte<M::Provenance>> {
        let Value::Variant { idx, data } = val else { panic!() };

        let mut bytes = list![AbstractByte::Uninit; size.bytes()];

>         // idx is to be guaranteed in bounds by the well-formed check in the typed store.

Reviewer: Please add a comment explaining why we write the data first

    fn encode<M: Memory>(Type::Enum { variants, .. }: Self, val: Value<M>) -> List<AbstractByte<M::Provenance>> {
        let Value::Variant { idx, data } = val else { panic!() };

        // `idx` is guaranteed to be in bounds by the well-formed check in the type.
        let Variant { ty: variant, tagger } = variants[idx];
        let mut bytes = variant.encode(data.extract());

        // Write tag into the bytes around the data.
        // This is fine as we don't allow encoded data and the tag to overlap.
        for (offset, value) in tagger.iter() {

>             assert_eq!(Some(AbstractByte::Uninit), bytes.get(offset.bytes()));

Author: For some reason it doesn't compile when I use `bytes[offset.bytes()]` even though I am indexing in a `List` using an `Int`

Author: Otherwise this PR is ready for review again.

Reviewer: > doesn't compile

What's the error message? Don't just say "it doesn't work", always explain what happens instead. :)

Author: "cannot index into a value of type `libspecr::List<mem::AbstractByte<<M as mem::Memory>::Provenance>>`".
Reading further into this, apparently specr doesn't transpile correctly in the case of `assert_eq!(AbstractByte::Uninit, bytes[offset.bytes()]);`. I noted this for myself to maybe fix it in the future, should I make an issue in the minirust-tooling repo?

Reviewer: Ah, is the macro getting in the way? That's possible. 

Sure, please file an issue! This will likely be non-trivial to fix though, and it's not very important.

Author: Changed everything else as requested.
